# MAAGAP — Objectives 1–4: Multi-Stage Predictive Framework

**Machine Analytics for Allocation, Governance and Assessment of Projects**

This notebook implements and evaluates all four study objectives:

| Objective | Description |
|-----------|-------------|
| **1** | Multi-stage predictive framework (RF + XGBoost + LSTM + Meta-ensemble) |
| **2** | Model evaluation — Accuracy, Precision, Recall, F1-Score, AUC-ROC, MAE |
| **3** | Dynamic Risk Scoring Engine — translates probabilities into Low/Medium/High/Critical tiers |
| **4** | LP Resource Allocation Optimization — ≥15% efficiency improvement over manual baseline |

**Data sources:**
- PPDO Iloilo 2026 project monitoring records (`MONITORING REPORT Con`)
- Fund Transfer Con sheet (21k+ real fund transfer records, 2013–2026)
- Synthetic dataset: 3,000 projects (2016–2025) grounded in real distributions
- External: PAGASA weather data & PSA economic indicators (CPI, CMRPI)


In [1]:
import os, sys, warnings
warnings.filterwarnings('ignore')
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

import numpy as np
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.metrics import roc_curve

from maagap.config import SEED, OUTPUTS_DIR, RISK_LABELS, RISK_THRESHOLDS
from maagap.data_preprocessing import (
    load_and_clean_ppdo, extract_distributions,
    load_fund_transfer_con, extract_fund_transfer_distributions,
)
from maagap.synthetic_generator import generate_synthetic_dataset
from maagap.feature_engineering import (
    build_static_features, build_temporal_sequences,
    build_targets, split_data,
)
from maagap.models import (
    train_random_forest, train_xgboost, train_lstm,
    train_meta_ensemble, predict_meta, meta_ensemble_percent_contributions,
)
from maagap.evaluation import (
    binary_metrics, regression_metrics, multiclass_metrics,
    plot_confusion_matrix, plot_roc_curves, plot_feature_importance,
    plot_training_history, plot_risk_distribution, plot_risk_distribution_rf_xgb,
    plot_model_comparison, generate_full_report,
    # Objective 3
    plot_risk_score_distribution, plot_risk_tier_comparison, plot_logic_consistency,
    # Objective 4
    plot_optimization_comparison, plot_lp_selection_profile,
)
from maagap.risk_scoring import (
    compute_risk_score, risk_tiers, logic_consistency_check, RiskScoringConfig,
)

np.random.seed(SEED)
print("All imports OK")
print(f"RISK_THRESHOLDS: {RISK_THRESHOLDS}")


All imports OK
RISK_THRESHOLDS: {'Low': (0.0, 0.3), 'Medium': (0.3, 0.7), 'High': (0.7, 0.9), 'Critical': (0.9, 1.0)}


## Step 1 — Load & Clean Real PPDO Data + Fund Transfer Con

We load **two real data sources**:

1. **`MONITORING REPORT Con`** — the PPDO 2026 project list used in Objective 1 to extract
   budget log-normal parameters, agency distribution, and project-type ratios.
2. **`Fund Transfer Con`** — 21k+ real fund transfer records (2013–2026) providing supplementary
   municipality distribution, funding-source distribution, liquidation rates, and a larger
   budget sample. **Previously unused — now integrated.**


In [2]:
# ============================================================
# STEP 1: LOAD & CLEAN REAL DATA (PPDO & FUND TRANSFER)
# ============================================================

# --- MONITORING REPORT Con ---
ppdo_df = load_and_clean_ppdo()
distributions = extract_distributions(ppdo_df)

print(f"MONITORING REPORT Con — Cleaned records : {len(ppdo_df)}")
print(f"  Budget log-mean : {distributions['budget_log_mean']:.2f}")
print(f"  Budget log-std  : {distributions['budget_log_std']:.2f}")
print("  Type distribution:")
for k, v in distributions["type_probs"].items():
    print(f"    {k}: {v:.1%}")

# --- Fund Transfer Con supplementary data ---
print()
df_ft = load_fund_transfer_con()
ft_dist = {}
if df_ft is not None:
    ft_dist = extract_fund_transfer_distributions(df_ft)
    print(f"Fund Transfer Con — Records loaded    : {len(df_ft):,}")
    print(f"Fund Transfer Con — Year range        : {int(df_ft['year'].min())} – {int(df_ft['year'].max())}")
    print(f"Fund Transfer Con — Budget log-mean  : {ft_dist.get('ft_budget_log_mean', 0):.2f}")
    print(f"Fund Transfer Con — Budget median    : PHP {ft_dist.get('ft_budget_median', 0):,.0f}")
    print(f"Fund Transfer Con — Liquidation rate : {ft_dist.get('liquidation_rate', 0):.1%}")
    print(f"Fund Transfer Con — Unliquidated rate: {ft_dist.get('unliquidated_rate', 0):.1%}")
    top_munis = sorted(ft_dist.get("municipality_probs", {}).items(), key=lambda x: -x[1])[:5]
    print("Fund Transfer Con — Top municipalities:")
    for m, p in top_munis:
        print(f"  {m}: {p:.1%}")
    distributions["ft_distributions"] = ft_dist
else:
    print("Fund Transfer Con — Sheet not available (skipped)")


MONITORING REPORT Con — Cleaned records : 8633
  Budget log-mean : 11.29
  Budget log-std  : 1.53
  Type distribution:
    Non-Infrastructure: 74.3%
    Infrastructure: 25.7%

Fund Transfer Con — Records loaded    : 21,083
Fund Transfer Con — Year range        : 2013 – 2026
Fund Transfer Con — Budget log-mean  : 11.03
Fund Transfer Con — Budget median    : PHP 50,000
Fund Transfer Con — Liquidation rate : 95.0%
Fund Transfer Con — Unliquidated rate: 2.6%
Fund Transfer Con — Top municipalities:
  Dumangas: 6.1%
  Sta.Barbara: 5.5%
  Lambunao: 5.2%
  Leon: 5.2%
  Calinog: 4.9%


### Visualise Real Data

In [3]:
budget_data = ppdo_df['approved_budget'].dropna()
budget_data = budget_data[budget_data > 0]

fig = make_subplots(rows=1, cols=3, subplot_titles=(
    'Budget Distribution (log scale)', 'Project Type', 'Status'
))

fig.add_trace(go.Histogram(
    x=np.log10(budget_data), nbinsx=40, marker_color='#3498db', name='Budget',
    hovertemplate='Log10(Budget): %{x:.1f}<br>Count: %{y}<extra></extra>',
), row=1, col=1)

type_vc = ppdo_df['project_type'].value_counts()
fig.add_trace(go.Bar(
    x=type_vc.index, y=type_vc.values,
    marker_color=['#2ecc71', '#e74c3c'], name='Type',
    text=type_vc.values, textposition='auto',
), row=1, col=2)

status_vc = ppdo_df['status'].value_counts()
fig.add_trace(go.Bar(
    x=status_vc.index, y=status_vc.values,
    marker_color='#9b59b6', name='Status',
    text=status_vc.values, textposition='auto',
), row=1, col=3)

fig.update_layout(
    title='Real PPDO 2026 Data Overview', showlegend=False,
    template='plotly_white', height=400, width=1100,
)
fig.show()

print(f'\nSample rows:')
ppdo_df[['project_name', 'implementing_agency', 'approved_budget', 'project_type', 'status']].head(10)


Sample rows:


,project_name,implementing_agency,approved_budget,project_type,status
0,Rehabilitation/Extension of Streetlights,Municipality,100000.0,Infrastructure,completed
1,Rehabilitation of Streetlights,Brgy.,30000.0,Infrastructure,completed
2,Streetlights,Brgy.,30000.0,Non-Infrastructure,completed
3,Rehabilitation of Multi-Purpose Center,Brgy.,15000.0,Infrastructure,completed
4,Purchase of 1 unit computer with printer with ...,Municipality,40000.0,Non-Infrastructure,completed
5,School Building Magbato-Sibacungan Elem. School,Municipality,300000.0,Infrastructure,completed
6,Purchase of Lot of Brgy. Coto for Infrastructu...,Municipality,150000.0,Infrastructure,completed
7,Improvement of Coto Water System,Brgy.,45000.0,Infrastructure,completed
8,"Multi-Purpose Hall of Brgy. Coto, Lambunao",Municipality,300000.0,Non-Infrastructure,ongoing
9,Improvement of Multi-Purpose Hall,Brgy.,30000.0,Infrastructure,completed


## Step 2 — Generate Synthetic Dataset (2016–2025)

3,000 synthetic projects are generated using distributions extracted from the real PPDO + Fund Transfer
data. Each project includes:
- Project metadata (agency, type, budget, location, procurement mode, funding source)
- PAGASA weather exposure (rainfall, typhoon days) per quarter
- PSA economic indicators (CPI, CMRPI) per quarter
- Simulated outcomes: `is_delayed`, `delay_days`, `is_cost_overrun`, `risk_category`


In [4]:
# ============================================================
# STEP 2: GENERATE MULTI-YEAR SYNTHETIC DATASET
# ============================================================

df_proj, df_qtr = generate_synthetic_dataset(distributions)

print(f'Projects generated : {len(df_proj)}')
print(f'Quarterly records  : {len(df_qtr)}')
print(f'Delayed projects   : {df_proj["is_delayed"].sum()} ({df_proj["is_delayed"].mean():.1%})')
print(f'Cost-overrun projects: {df_proj["is_cost_overrun"].sum()} ({df_proj["is_cost_overrun"].mean():.1%})')
print(f'\nRisk Category Distribution:')
for cat in RISK_LABELS:
    n = (df_proj['risk_category'] == cat).sum()
    print(f'  {cat}: {n} ({n/len(df_proj):.1%})')

Projects generated : 3000
Quarterly records  : 7944
Delayed projects   : 493 (16.4%)
Cost-overrun projects: 1149 (38.3%)

Risk Category Distribution:
  Low: 2003 (66.8%)
  Medium: 845 (28.2%)
  High: 148 (4.9%)
  Critical: 4 (0.1%)


### Synthetic Data Visualisations

In [5]:
fig = make_subplots(
    rows=2, cols=3,
    subplot_titles=(
        'Risk Category Distribution', 'Delay Probability Distribution',
        'Budget by Project Type', 'Delay Rate by Agency',
        'Sample Project Progression', 'Average Weather by Quarter',
    ),
    vertical_spacing=0.12, horizontal_spacing=0.08,
)

risk_colors = {'Low': '#2ecc71', 'Medium': '#f39c12', 'High': '#e74c3c', 'Critical': '#8e44ad'}
risk_vc = df_proj['risk_category'].value_counts().reindex(RISK_LABELS)
fig.add_trace(go.Bar(
    x=risk_vc.index, y=risk_vc.values,
    marker_color=[risk_colors[r] for r in risk_vc.index],
    text=risk_vc.values, textposition='auto', showlegend=False,
), row=1, col=1)

fig.add_trace(go.Histogram(
    x=df_proj['delay_probability'], nbinsx=50,
    marker_color='#e74c3c', showlegend=False,
), row=1, col=2)

for ptype, color in [('Infrastructure', '#e74c3c'), ('Non-Infrastructure', '#3498db')]:
    subset = df_proj[df_proj['project_type'] == ptype]
    fig.add_trace(go.Histogram(
        x=np.log10(subset['approved_budget']), name=ptype,
        marker_color=color, opacity=0.7, nbinsx=30,
    ), row=1, col=3)

agency_delay = df_proj.groupby('implementing_agency')['is_delayed'].mean().sort_values(ascending=True)
fig.add_trace(go.Bar(
    x=agency_delay.values, y=agency_delay.index, orientation='h',
    marker_color='#e67e22', showlegend=False,
), row=2, col=1)

sample_proj = df_proj.iloc[0]
sample_qtr = df_qtr[df_qtr['project_id'] == sample_proj['project_id']].sort_values('quarter')
fig.add_trace(go.Scatter(
    x=sample_qtr['quarter'], y=sample_qtr['planned_progress_pct'],
    mode='lines+markers', name='Planned', line=dict(color='#2ecc71', dash='dash'),
), row=2, col=2)
fig.add_trace(go.Scatter(
    x=sample_qtr['quarter'], y=sample_qtr['actual_progress_pct'],
    mode='lines+markers', name='Actual', line=dict(color='#e74c3c'),
), row=2, col=2)

avg_weather = df_qtr.groupby('quarter')[['rainfall_mm', 'typhoon_days']].mean()
fig.add_trace(go.Bar(
    x=avg_weather.index, y=avg_weather['rainfall_mm'],
    name='Rainfall (mm)', marker_color='#3498db',
), row=2, col=3)
fig.add_trace(go.Scatter(
    x=avg_weather.index, y=avg_weather['typhoon_days'] * 50,
    name='Typhoon Days (Ã - 50)', mode='lines+markers', line=dict(color='#e74c3c'),
), row=2, col=3)

fig.update_layout(
    height=750, width=1200, template='plotly_white',
    title='Synthetic Dataset Overview',
    showlegend=True,
)
fig.show()

## Step 3 Feature Engineering

We transform the raw data into ML-ready formats:
- **Static features** (30 columns): numeric fields + engineered interaction terms + label-encoded categoricals
- **Temporal tensor** (3000 Ã -  4 Ã -  9): MinMax-scaled quarterly sequences for LSTM

In [6]:
# ============================================================
# STEP 3: FEATURE ENGINEERING (STATIC & TEMPORAL)
# ============================================================

X_static, feat_names, static_scaler, _ = build_static_features(df_proj)
X_temporal, temp_feat_names, temp_scaler = build_temporal_sequences(df_proj, df_qtr)
y_delay, y_overrun, y_risk, y_delay_days, y_overrun_pct = build_targets(df_proj)

print(f'Static feature matrix : {X_static.shape}')
print(f'Temporal tensor       : {X_temporal.shape}')
print(f'\nStatic feature names ({len(feat_names)}):')
for i, name in enumerate(feat_names):
    print(f'  {i+1:2d}. {name}')
print(f'\nTemporal features ({len(temp_feat_names)}): {temp_feat_names}')
print(f'\nTarget distributions:')
print(f'  Delay: 0={np.sum(y_delay==0)}, 1={np.sum(y_delay==1)}')
print(f'  Risk:  ' + ', '.join(f'{RISK_LABELS[i]}={np.sum(y_risk==i)}' for i in range(len(RISK_LABELS))))

Static feature matrix : (3000, 30)
Temporal tensor       : (3000, 4, 9)

Static feature names (30):
   1. approved_budget
   2. planned_duration_months
   3. start_month
   4. has_contractor
   5. contractor_reliability
   6. agency_capacity
   7. typhoon_exposure
   8. cpi_at_start
   9. cmrpi_at_start
  10. cpi_change
  11. cmrpi_change
  12. budget_log
  13. is_infrastructure
  14. is_typhoon_start
  15. infra_x_typhoon
  16. infra_x_budget
  17. contractor_x_typhoon
  18. budget_x_cpi_change
  19. low_contractor_flag
  20. high_budget_flag
  21. agency_risk
  22. contractor_x_agency
  23. infra_x_low_contractor
  24. typhoon_x_budget
  25. econ_pressure
  26. composite_risk_features
  27. project_type_enc
  28. implementing_agency_enc
  29. procurement_mode_enc
  30. funding_source_enc

Temporal features (9): ['planned_progress_pct', 'actual_progress_pct', 'slippage_pct', 'expenditure_ratio', 'issues_count', 'rainfall_mm', 'typhoon_days', 'cpi_quarterly', 'cmrpi_quarterly']

Target

## Step 4 — Data Split: **70 / 15 / 15**

| Partition | Size | Purpose |
|-----------|------|---------|
| **Train** (70%) | 2,100 projects | Model training |
| **Validation** (15%) | 450 projects | Hyperparameter tuning & early stopping |
| **Test** (15%) | 450 projects | Final evaluation (never seen during training) |

The validation set is used for LSTM early-stopping and LSTM/RF/XGB hyperparameter search.
The test set is held out completely and used only for reporting.


In [7]:
# ============================================================
# STEP 4: DATA SPLIT (70% TRAIN / 15% VAL / 15% TEST)
# ============================================================

idx_tr, idx_va, idx_te = split_data(len(df_proj))

n = len(df_proj)
print(f"Total projects : {n}")
print(f"Train          : {len(idx_tr)} ({len(idx_tr)/n:.1%})")
print(f"Validation     : {len(idx_va)} ({len(idx_va)/n:.1%})")
print(f"Test           : {len(idx_te)} ({len(idx_te)/n:.1%})")
print()

# Verify no overlap
assert len(set(idx_tr) & set(idx_va)) == 0, "Train/Val overlap!"
assert len(set(idx_tr) & set(idx_te)) == 0, "Train/Test overlap!"
assert len(set(idx_va) & set(idx_te)) == 0, "Val/Test overlap!"
print("No partition overlap — split is correct.")

Xs_tr, Xs_va, Xs_te = X_static[idx_tr],   X_static[idx_va],   X_static[idx_te]
Xt_tr, Xt_va, Xt_te = X_temporal[idx_tr], X_temporal[idx_va], X_temporal[idx_te]
yd_tr, yd_va, yd_te = y_delay[idx_tr],    y_delay[idx_va],    y_delay[idx_te]
yr_tr, yr_va, yr_te = y_risk[idx_tr],     y_risk[idx_va],     y_risk[idx_te]
ydd_te = y_delay_days[idx_te]


Total projects : 3000
Train          : 2100 (70.0%)
Validation     : 450 (15.0%)
Test           : 450 (15.0%)

No partition overlap — split is correct.


## Step 5 Train Models

### Stage 1a: Random Forest (with hyperparameter tuning)

### Baseline delay models + Meta-Ensemble (baseline bases)

We first train **untuned** RF/XGB/LSTM for delay, then fit a logistic-regression meta-learner on their validation probabilities. This is compared later with **meta on tuned bases** after `RandomizedSearchCV` / LSTM search.


In [8]:
print('Baseline delay models (no tuning) + meta-ensemble on baseline bases...')
rf_b = train_random_forest(Xs_tr, yd_tr, task='delay', tune=False)
rf_prob_va_b = rf_b.predict_proba(Xs_va)[:, 1]
rf_prob_te_b = rf_b.predict_proba(Xs_te)[:, 1]
xgb_b = train_xgboost(Xs_tr, yd_tr, task='delay', tune=False)
xgb_prob_va_b = xgb_b.predict_proba(Xs_va)[:, 1]
xgb_prob_te_b = xgb_b.predict_proba(Xs_te)[:, 1]
lstm_b, _, _ = train_lstm(Xt_tr, yd_tr, Xt_va, yd_va, task='delay', tune=False)
lstm_prob_va_b = lstm_b.predict(Xt_va, verbose=0).flatten()
lstm_prob_te_b = lstm_b.predict(Xt_te, verbose=0).flatten()
meta_b = train_meta_ensemble(
    rf_prob_va_b, xgb_prob_va_b, lstm_prob_va_b, yd_va,
    artifact_name='meta_ensemble_baseline.pkl',
)
meta_b_pred_te, meta_b_prob_te = predict_meta(
    meta_b, rf_prob_te_b, xgb_prob_te_b, lstm_prob_te_b,
)
meta_b_prob_pos = meta_b_prob_te[:, 1] if meta_b_prob_te.ndim > 1 else meta_b_prob_te
print('  -> meta_ensemble_baseline.pkl')


Baseline delay models (no tuning) + meta-ensemble on baseline bases...
  -> meta_ensemble_baseline.pkl


In [9]:
# ============================================================
# STEP 5A: TRAIN RANDOM FOREST
# ============================================================

print('Training Random Forest for delay prediction (with RandomizedSearchCV)...')
rf = train_random_forest(Xs_tr, yd_tr, task='delay', tune=True)
rf_pred_te = rf.predict(Xs_te)
rf_prob_te = rf.predict_proba(Xs_te)[:, 1]
rf_prob_va = rf.predict_proba(Xs_va)[:, 1]
print(f'  n_estimators={rf.n_estimators}, max_depth={rf.max_depth}')

Training Random Forest for delay prediction (with RandomizedSearchCV)...
    RF best params: {'n_estimators': 500, 'min_samples_split': 2, 'min_samples_leaf': 4, 'max_features': 'log2', 'max_depth': 10}
  n_estimators=500, max_depth=10


### Stage 1b: XGBoost (with hyperparameter tuning)

In [10]:
# ============================================================
# STEP 5B: TRAIN XGBOOST
# ============================================================

print('Training XGBoost for delay prediction (with RandomizedSearchCV)...')
xgb = train_xgboost(Xs_tr, yd_tr, task='delay', tune=True)
xgb_pred_te = xgb.predict(Xs_te)
xgb_prob_te = xgb.predict_proba(Xs_te)[:, 1]
xgb_prob_va = xgb.predict_proba(Xs_va)[:, 1]
print(f'  n_estimators={xgb.n_estimators}, max_depth={xgb.max_depth}, lr={xgb.learning_rate}')

Training XGBoost for delay prediction (with RandomizedSearchCV)...
    XGB best params: {'subsample': 0.7, 'reg_lambda': 2.0, 'reg_alpha': 1.0, 'n_estimators': 300, 'min_child_weight': 3, 'max_depth': 8, 'learning_rate': 0.01, 'colsample_bytree': 0.7}
  n_estimators=300, max_depth=8, lr=0.01


### Stage 1c-d: Risk Categorisation Models

In [11]:
# ============================================================
# STEP 5C: TRAIN RISK CATEGORISATION MODELS
# ============================================================

print('Training Random Forest for risk categorisation...')
rf_risk = train_random_forest(Xs_tr, yr_tr, task='risk', tune=True)
rf_risk_pred_te = rf_risk.predict(Xs_te)

print('\nTraining XGBoost for risk categorisation...')
xgb_risk = train_xgboost(Xs_tr, yr_tr, task='risk', tune=True)
xgb_risk_pred_te = xgb_risk.predict(Xs_te)

Training Random Forest for risk categorisation...
    RF best params: {'n_estimators': 500, 'min_samples_split': 2, 'min_samples_leaf': 4, 'max_features': 'log2', 'max_depth': 10}

Training XGBoost for risk categorisation...
    XGB best params: {'subsample': 1.0, 'reg_lambda': 0.5, 'reg_alpha': 0, 'n_estimators': 300, 'min_child_weight': 3, 'max_depth': 12, 'learning_rate': 0.05, 'colsample_bytree': 0.8}


### Stage 2: LSTM (Temporal Delay Prediction)

In [12]:
# ============================================================
# STEP 5D: TRAIN LSTM (TEMPORAL SEQUENCE MODEL)
# ============================================================

print('Training LSTM for temporal delay prediction (with hyperparameter search)...')
print(f'  Input shape: ({X_temporal.shape[1]}, {X_temporal.shape[2]})')
lstm, history, lstm_best_params = train_lstm(Xt_tr, yd_tr, Xt_va, yd_va, task='delay', tune=True)
lstm_prob_te = lstm.predict(Xt_te, verbose=0).flatten()
lstm_pred_te = (lstm_prob_te >= 0.5).astype(int)
lstm_prob_va = lstm.predict(Xt_va, verbose=0).flatten()
print(f'  Epochs trained: {len(history.history["loss"])}')
print(f'  Final train loss: {history.history["loss"][-1]:.4f}')
print(f'  Final val loss:   {history.history["val_loss"][-1]:.4f}')
print(f'  Best LSTM config: {lstm_best_params}')


Training LSTM for temporal delay prediction (with hyperparameter search)...
  Input shape: (4, 9)
    LSTM hyperparameter search (8 configurations)...
    [1/8] units=(128,64), dropout=0.35, lr=0.001, batch=32
           val_loss=0.2062, val_acc=0.9222
    [2/8] units=(64,32), dropout=0.3, lr=0.001, batch=32
           val_loss=0.2174, val_acc=0.9289
    [3/8] units=(128,64), dropout=0.25, lr=0.0005, batch=32
           val_loss=0.2382, val_acc=0.9289
    [4/8] units=(96,48), dropout=0.35, lr=0.0005, batch=64
           val_loss=0.2210, val_acc=0.9222
    [5/8] units=(128,64), dropout=0.4, lr=0.002, batch=64
           val_loss=0.1900, val_acc=0.9289
    [6/8] units=(64,32), dropout=0.2, lr=0.001, batch=64
           val_loss=0.2308, val_acc=0.9267
    [7/8] units=(96,48), dropout=0.3, lr=0.001, batch=32
           val_loss=0.2108, val_acc=0.9244
    [8/8] units=(128,32), dropout=0.35, lr=0.001, batch=32
           val_loss=0.2524, val_acc=0.9200
    LSTM best params: {'units_1': 128, 

### LSTM Training History

In [13]:
fig = plot_training_history(history, 'lstm_training_history.png')
fig.show()

### Meta-Ensemble (Stacking RF + XGBoost + LSTM)

In [14]:
# ============================================================
# STEP 5E: TRAIN META-ENSEMBLE (LOGISTIC STACKING)
# ============================================================

print('Training Meta-Ensemble on TUNED base models (stacking RF + XGB + LSTM)...')
meta = train_meta_ensemble(
    rf_prob_va, xgb_prob_va, lstm_prob_va, yd_va,
    artifact_name='meta_ensemble.pkl',
)
meta_pred_te, meta_prob_te = predict_meta(meta, rf_prob_te, xgb_prob_te, lstm_prob_te)
meta_prob_pos = meta_prob_te[:, 1] if meta_prob_te.ndim > 1 else meta_prob_te

from maagap.models import meta_ensemble_percent_contributions
pct = meta_ensemble_percent_contributions(meta, rf_prob_va, xgb_prob_va, lstm_prob_va)
print(f'  Meta-ensemble contribution (%): {pct}')
print(f'  Raw stacking coefficients: {meta.coef_[0]}')


Training Meta-Ensemble on TUNED base models (stacking RF + XGB + LSTM)...
  Meta-ensemble contribution (%): {'Random Forest': 14.74, 'XGBoost': 12.75, 'LSTM': 72.51}
  Raw stacking coefficients: [1.14117757 0.9618077  4.64074508]


## Objective 3 — Dynamic Risk Scoring Engine

This section implements the manuscript's third objective:
> *"Design a dynamic risk scoring engine that translates probability outputs from the predictive models
> into actionable tiers — Low, Medium, High, and Critical — by applying logic consistency testing
> and defined threshold boundaries to ensure the reliability of risk classifications."*

### Scoring Formula
$$\text{risk\_score} = 0.55 \times P(\text{delay}) + 0.45 \times P(\text{overrun})$$

### Threshold Boundaries (from manuscript)
| Tier | Score Range |
|------|------------|
| Low | [0.0, 0.30) |
| Medium | [0.30, 0.70) |
| High | [0.70, 0.90) |
| Critical | [0.90, 1.0] |


In [15]:
# ============================================================
# OBJECTIVE 3: DYNAMIC RISK SCORING ENGINE
# ============================================================

# --- Objective 3: Dynamic Risk Scoring Engine ---

# Inputs:
#   delay_prob   = tuned meta-ensemble probability output (P(delay))
#   overrun_prob = synthetic overrun probability (from generation parameters)
overrun_proba_te = df_proj["overrun_probability"].values[idx_te]
risk_cfg         = RiskScoringConfig(w_delay=0.55, w_overrun=0.45)
risk_scores_te   = compute_risk_score(meta_prob_pos, overrun_proba_te, cfg=risk_cfg)
risk_tiers_te    = risk_tiers(risk_scores_te)
actual_tiers_te  = np.array([RISK_LABELS[i] for i in yr_te])

# Logic Consistency Check
consistency = logic_consistency_check(risk_scores_te, risk_tiers_te)
print(f"Logic consistency violations : {consistency['violations']} / {len(risk_scores_te)}")
print(f"Unknown tier labels          : {consistency['unknown_tier_labels']}")
if consistency['violations'] == 0:
    print("=> PASS: All tier assignments are consistent with threshold boundaries.")

# Tier distribution
print("\nRisk Tier Distribution (Test Set):")
print(f"{'Tier':10s} {'N':>6s} {'%':>7s}  {'Score Range':>15s}  {'Mean Score':>12s}  {'Std Score':>10s}")
print("-" * 72)
for tier in RISK_LABELS:
    mask    = (risk_tiers_te == tier)
    cnt     = mask.sum()
    lo, hi  = RISK_THRESHOLDS[tier]
    sc      = risk_scores_te[mask]
    mean_s  = sc.mean() if cnt > 0 else 0.0
    std_s   = sc.std()  if cnt > 0 else 0.0
    print(f"{tier:10s} {cnt:>6d} {cnt/len(risk_tiers_te):>7.1%}  [{lo:.2f}, {hi:.2f})       {mean_s:>12.4f}  {std_s:>10.4f}")

# Agreement with ground-truth risk categories
tier_match = (risk_tiers_te == actual_tiers_te)
print(f"\nTier agreement with ground truth : {tier_match.sum()} / {len(tier_match)} ({tier_match.mean():.1%})")

# Per-tier breakdown
print("\nPer-Tier Agreement:")
print(f"{'Tier':10s} {'Predicted':>10s} {'Actual':>8s} {'Match':>8s} {'Precision':>10s}")
print("-" * 52)
for tier in RISK_LABELS:
    pred_m   = (risk_tiers_te   == tier)
    actual_m = (actual_tiers_te == tier)
    match    = np.sum(pred_m & actual_m)
    prec     = match / max(pred_m.sum(), 1)
    print(f"{tier:10s} {pred_m.sum():>10d} {actual_m.sum():>8d} {match:>8d} {prec:>10.1%}")


Logic consistency violations : 0 / 450
Unknown tier labels          : 0
=> PASS: All tier assignments are consistent with threshold boundaries.

Risk Tier Distribution (Test Set):
Tier            N       %      Score Range    Mean Score   Std Score
------------------------------------------------------------------------
Low           336   74.7%  [0.00, 0.30)             0.1474      0.0450
Medium         74   16.4%  [0.30, 0.70)             0.4798      0.1123
High           28    6.2%  [0.70, 0.90)             0.8513      0.0391
Critical       12    2.7%  [0.90, 1.00)             0.9148      0.0075

Tier agreement with ground truth : 360 / 450 (80.0%)

Per-Tier Agreement:
Tier        Predicted   Actual    Match  Precision
----------------------------------------------------
Low               336      297      288      85.7%
Medium             74      132       63      85.1%
High               28       21        9      32.1%
Critical           12        0        0       0.0%


In [16]:
# Objective 3 Visualisations

# 1. Risk score histogram colored by tier
fig1 = plot_risk_score_distribution(risk_scores_te, risk_tiers_te, "risk_score_distribution.png")
fig1.show()

# 2. Actual vs Predicted tier bar chart
fig2 = plot_risk_tier_comparison(actual_tiers_te, risk_tiers_te, "risk_tier_comparison.png")
fig2.show()

# 3. Logic consistency gauge
fig3 = plot_logic_consistency(consistency, len(risk_scores_te), "logic_consistency.png")
fig3.show()


## Objective 4 — LP Resource Allocation Optimization

This section implements the manuscript's fourth objective:
> *"Develop and evaluate optimization algorithms using linear programming to generate resource
> reallocation recommendations under constraints such as budget, manpower, and equipment.
> Success will be measured by demonstrating at least a 15% improvement in allocation efficiency
> compared to current manual approaches through simulated project scenarios."*

### Problem Formulation

Given a portfolio of **n = 450 test projects** and **limited inspection capacity**:
- **Decision variable**: $x_j \in \{0, 1\}$ — whether project $j$ is selected for inspection
- **Objective**: Maximise $\sum_j x_j \cdot u_j$ where $u_j$ is a weighted utility score
- **Constraint**: $\sum_j x_j \leq C$ (capacity = inspectors × inspections/period)

**Utility function:**
$$u_j = 0.65 \times \text{risk\_score}_j + 0.25 \times \frac{\text{tier\_weight}_j}{4} + 0.10 \times \text{importance}_j$$

**Capacity scenario** (per manuscript delimitation):
- 6 permanent inspectors × 5 inspections per planning period = **30 slots**

### Evaluation
| Metric | Description |
|--------|-------------|
| Efficiency | Average weighted risk captured per selected project |
| Improvement | (LP efficiency − Baseline efficiency) / Baseline × 100% |
| Target | ≥ 15% improvement |
| Validation | 200-iteration Monte Carlo simulation with Gaussian noise (σ = 0.05) |


In [17]:
# ============================================================
# OBJECTIVE 4: LP RESOURCE ALLOCATION OPTIMIZATION
# ============================================================

from maagap.optimization import (
    OptimizationConfig, optimize_inspection_allocation,
    baseline_manual_allocation, allocation_efficiency,
    compute_efficiency_improvement, monte_carlo_robustness,
)

opt_cfg  = OptimizationConfig(inspectors_available=6, inspections_per_inspector=5, seed=SEED)
capacity = opt_cfg.inspectors_available * opt_cfg.inspections_per_inspector
n_test   = len(risk_scores_te)

print(f"Test-set projects   : {n_test}")
print(f"Inspection capacity : {opt_cfg.inspectors_available} inspectors × "
      f"{opt_cfg.inspections_per_inspector} = {capacity} slots")

# Baseline (manual / random) allocation
result_base  = baseline_manual_allocation(risk_scores_te, cfg=opt_cfg)
eff_base_val = allocation_efficiency(result_base["selected_idx"], risk_scores_te, risk_tiers_te)["avg_captured_risk"]

# LP optimised allocation
result_lp   = optimize_inspection_allocation(risk_scores_te, risk_tiers_te, cfg=opt_cfg)
eff_lp_val  = allocation_efficiency(result_lp["selected_idx"],  risk_scores_te, risk_tiers_te)["avg_captured_risk"]

# Efficiency improvement
eff_summary = compute_efficiency_improvement(
    result_lp["selected_idx"], result_base["selected_idx"], risk_scores_te, risk_tiers_te
)
improvement_pct = eff_summary["improvement_pct"]

print(f"\n{'Method':32s} {'Selected':>10s} {'Avg Risk Score':>16s}")
print("-" * 62)
print(f"{'Baseline (Manual/Random)':32s} {result_base['selected_count']:>10d} {eff_base_val:>16.4f}")
print(f"{'LP Optimized':32s} {result_lp['selected_count']:>10d} {eff_lp_val:>16.4f}")
target_label = "PASS (>= 15%)" if eff_summary["target_met"] else "FAIL (< 15%)"
print(f"{'Efficiency Improvement':32s} {'':>10s} {improvement_pct:>15.2f}%  [{target_label}]")
print(f"LP Solver Status : {result_lp['status']}")

# Coverage by risk tier
print("\nCoverage by Risk Tier — Inspected Projects:")
print(f"{'Tier':10s} {'Total':>8s} {'Baseline':>10s} {'LP Opt':>10s} {'LP Recall':>12s}")
print("-" * 55)
base_sel_set = set(result_base["selected_idx"].tolist())
lp_sel_set   = set(result_lp["selected_idx"].tolist())
for tier in RISK_LABELS:
    tier_idx  = np.where(risk_tiers_te == tier)[0]
    n_total   = len(tier_idx)
    n_base    = sum(1 for i in tier_idx if i in base_sel_set)
    n_lp      = sum(1 for i in tier_idx if i in lp_sel_set)
    lp_recall = n_lp / max(n_total, 1)
    print(f"{tier:10s} {n_total:>8d} {n_base:>10d} {n_lp:>10d} {lp_recall:>11.1%}")


Test-set projects   : 450
Inspection capacity : 6 inspectors × 5 = 30 slots

Method                             Selected   Avg Risk Score
--------------------------------------------------------------
Baseline (Manual/Random)                 30           0.2165
LP Optimized                             30           0.7933
Efficiency Improvement                               266.34%  [PASS (>= 15%)]
LP Solver Status : Optimal

Coverage by Risk Tier — Inspected Projects:
Tier          Total   Baseline     LP Opt    LP Recall
-------------------------------------------------------
Low             336         25          0        0.0%
Medium           74          4          0        0.0%
High             28          0         18       64.3%
Critical         12          1         12      100.0%


In [18]:
from maagap.optimization import (
    OptimizationConfig, optimize_inspection_allocation,
    baseline_manual_allocation, allocation_efficiency,
    compute_efficiency_improvement, monte_carlo_robustness,
)

opt_cfg  = OptimizationConfig(inspectors_available=6, inspections_per_inspector=5, seed=SEED)
capacity = opt_cfg.inspectors_available * opt_cfg.inspections_per_inspector
n_test   = len(risk_scores_te)

print(f"Test-set projects   : {n_test}")
print(f"Inspection capacity : {opt_cfg.inspectors_available} inspectors × "
      f"{opt_cfg.inspections_per_inspector} = {capacity} slots")

# Baseline (manual / random) allocation
result_base  = baseline_manual_allocation(risk_scores_te, cfg=opt_cfg)
eff_base_val = allocation_efficiency(result_base["selected_idx"], risk_scores_te, risk_tiers_te)["avg_captured_risk"]

# LP optimised allocation
result_lp   = optimize_inspection_allocation(risk_scores_te, risk_tiers_te, cfg=opt_cfg)
eff_lp_val  = allocation_efficiency(result_lp["selected_idx"],  risk_scores_te, risk_tiers_te)["avg_captured_risk"]

# Efficiency improvement
eff_summary = compute_efficiency_improvement(
    result_lp["selected_idx"], result_base["selected_idx"], risk_scores_te, risk_tiers_te
)
improvement_pct = eff_summary["improvement_pct"]

print(f"\n{'Method':32s} {'Selected':>10s} {'Avg Risk Score':>16s}")
print("-" * 62)
print(f"{'Baseline (Manual/Random)':32s} {result_base['selected_count']:>10d} {eff_base_val:>16.4f}")
print(f"{'LP Optimized':32s} {result_lp['selected_count']:>10d} {eff_lp_val:>16.4f}")
target_label = "PASS (>= 15%)" if eff_summary["target_met"] else "FAIL (< 15%)"
print(f"{'Efficiency Improvement':32s} {'':>10s} {improvement_pct:>15.2f}%  [{target_label}]")
print(f"LP Solver Status : {result_lp['status']}")

# Coverage by risk tier
print("\nCoverage by Risk Tier — Inspected Projects:")
print(f"{'Tier':10s} {'Total':>8s} {'Baseline':>10s} {'LP Opt':>10s} {'LP Recall':>12s}")
print("-" * 55)
base_sel_set = set(result_base["selected_idx"].tolist())
lp_sel_set   = set(result_lp["selected_idx"].tolist())
for tier in RISK_LABELS:
    tier_idx  = np.where(risk_tiers_te == tier)[0]
    n_total   = len(tier_idx)
    n_base    = sum(1 for i in tier_idx if i in base_sel_set)
    n_lp      = sum(1 for i in tier_idx if i in lp_sel_set)
    lp_recall = n_lp / max(n_total, 1)
    print(f"{tier:10s} {n_total:>8d} {n_base:>10d} {n_lp:>10d} {lp_recall:>11.1%}")


Test-set projects   : 450
Inspection capacity : 6 inspectors × 5 = 30 slots

Method                             Selected   Avg Risk Score
--------------------------------------------------------------
Baseline (Manual/Random)                 30           0.2165
LP Optimized                             30           0.7933
Efficiency Improvement                               266.34%  [PASS (>= 15%)]
LP Solver Status : Optimal

Coverage by Risk Tier — Inspected Projects:
Tier          Total   Baseline     LP Opt    LP Recall
-------------------------------------------------------
Low             336         25          0        0.0%
Medium           74          4          0        0.0%
High             28          0         18       64.3%
Critical         12          1         12      100.0%


In [19]:
# ============================================================
# OBJECTIVE 4: MONTE CARLO ROBUSTNESS SIMULATION
# ============================================================

# Monte Carlo Robustness Simulation
print("Running Monte Carlo Robustness (200 iterations, noise_std=0.05)...")
mc_results = monte_carlo_robustness(
    risk_scores_te, risk_tiers_te,
    n_simulations=200, noise_std=0.05, cfg=opt_cfg,
)
print(f"  Successful runs     : {mc_results['n_successful']} / {mc_results['n_simulations']}")
print(f"  Mean improvement    : {mc_results['mean_improvement_pct']:.2f}%")
print(f"  Std dev             : {mc_results['std_improvement_pct']:.2f}%")
print(f"  Range               : [{mc_results['min_improvement_pct']:.2f}%, "
      f"{mc_results['max_improvement_pct']:.2f}%]")
print(f"  Simulations >= 15%  : {mc_results['pct_above_15']:.1f}%")

robust_label = "ROBUST" if mc_results['pct_above_15'] >= 90.0 else "CHECK"
print(f"\n=> {robust_label}: LP outperforms baseline by >= 15% in "
      f"{mc_results['pct_above_15']:.0f}% of Monte Carlo simulations.")


Running Monte Carlo Robustness (200 iterations, noise_std=0.05)...
  Successful runs     : 200 / 200
  Mean improvement    : 267.48%
  Std dev             : 11.81%
  Range               : [240.13%, 301.64%]
  Simulations >= 15%  : 100.0%

=> ROBUST: LP outperforms baseline by >= 15% in 100% of Monte Carlo simulations.


In [20]:
# Objective 4 Visualisations

# 1. Baseline vs LP efficiency comparison + MC distribution
fig4 = plot_optimization_comparison(
    eff_base_val, eff_lp_val, improvement_pct, mc_results,
    "optimization_comparison.png",
)
fig4.show()

# 2. LP selection profile: which projects each method chose
fig5 = plot_lp_selection_profile(
    risk_scores_te, risk_tiers_te,
    result_lp["selected_idx"], result_base["selected_idx"],
    "lp_selection_profile.png",
)
fig5.show()


### Hyperparameter Tuning Comparison

We **reuse** the untuned delay models from Step 5 (`rf_b`, `xgb_b`, `lstm_b`) — the same checkpoints used for **Meta (baseline bases)** — and compare them on the test set against the **tuned** RF, XGBoost, and LSTM. This matches `python main.py` and avoids training a second, slightly different “default” model.

In [21]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

print("Reusing baseline delay models from Step 5 (rf_b, xgb_b, lstm_b)...")

def _calc_metrics(y_true, y_pred, y_prob, label):
    return {
        "Model": label,
        "Accuracy": accuracy_score(y_true, y_pred),
        "Precision": precision_score(y_true, y_pred, zero_division=0),
        "Recall": recall_score(y_true, y_pred, zero_division=0),
        "F1-Score": f1_score(y_true, y_pred, zero_division=0),
        "AUC-ROC": roc_auc_score(y_true, y_prob),
    }

rf_bl_p = rf_b.predict(Xs_te)
rf_bl_pr = rf_b.predict_proba(Xs_te)[:, 1]
xgb_bl_p = xgb_b.predict(Xs_te)
xgb_bl_pr = xgb_b.predict_proba(Xs_te)[:, 1]
lstm_bl_pr = lstm_b.predict(Xt_te, verbose=0).flatten()
lstm_bl_p = (lstm_bl_pr >= 0.5).astype(int)

rf_bl_metrics = _calc_metrics(yd_te, rf_bl_p, rf_bl_pr, "RF (Default)")
rf_tuned_metrics = _calc_metrics(yd_te, rf_pred_te, rf_prob_te, "RF (Tuned)")
xgb_bl_metrics = _calc_metrics(yd_te, xgb_bl_p, xgb_bl_pr, "XGB (Default)")
xgb_tuned_metrics = _calc_metrics(yd_te, xgb_pred_te, xgb_prob_te, "XGB (Tuned)")
lstm_bl_metrics = _calc_metrics(yd_te, lstm_bl_p, lstm_bl_pr, "LSTM (Default)")
lstm_tuned_metrics = _calc_metrics(yd_te, lstm_pred_te, lstm_prob_te, "LSTM (Tuned)")

comparison = pd.DataFrame([
    rf_bl_metrics, rf_tuned_metrics,
    xgb_bl_metrics, xgb_tuned_metrics,
    lstm_bl_metrics, lstm_tuned_metrics,
])
comparison = comparison.set_index("Model")

print("\n--- Hyperparameter Tuning Comparison (test set) ---\n")
display(comparison.style.format("{:.4f}").background_gradient(cmap="Greens", axis=0))

keys = ["Accuracy", "Precision", "Recall", "F1-Score", "AUC-ROC"]
for model in ["RF", "XGB", "LSTM"]:
    bl = comparison.loc[f"{model} (Default)"]
    tu = comparison.loc[f"{model} (Tuned)"]
    diff = tu - bl
    print(f"\n{model} improvement after tuning:")
    for col in keys:
        arrow = "\u2191" if diff[col] > 0 else "\u2193" if diff[col] < 0 else "\u2194"
        print(f"  {col:12s}: {diff[col]:+.4f} {arrow}")

print("\nKey hyperparameters found by RandomizedSearchCV:")
print(f"  RF Tuned:  n_estimators={rf.n_estimators}, max_depth={rf.max_depth}, "
      f"min_samples_split={rf.min_samples_split}, min_samples_leaf={rf.min_samples_leaf}, "
      f"max_features={rf.max_features}")
print(f"  XGB Tuned: n_estimators={xgb.n_estimators}, max_depth={xgb.max_depth}, "
      f"lr={xgb.learning_rate}, subsample={xgb.subsample}, "
      f"colsample_bytree={xgb.colsample_bytree}")


Reusing baseline delay models from Step 5 (rf_b, xgb_b, lstm_b)...

--- Hyperparameter Tuning Comparison (test set) ---



,Accuracy,Precision,Recall,F1-Score,AUC-ROC
Model,,,,,
RF (Default),0.8622,0.5682,0.3676,0.4464,0.8325
RF (Tuned),0.8400,0.4767,0.6029,0.5325,0.8430
XGB (Default),0.8533,0.5179,0.4265,0.4677,0.8184
XGB (Tuned),0.8356,0.4694,0.6765,0.5542,0.8424
LSTM (Default),0.8622,0.5366,0.6471,0.5867,0.9159
LSTM (Tuned),0.8689,0.5584,0.6324,0.5931,0.9062



RF improvement after tuning:
  Accuracy    : -0.0222 ↓
  Precision   : -0.0914 ↓
  Recall      : +0.2353 ↑
  F1-Score    : +0.0860 ↑
  AUC-ROC     : +0.0105 ↑

XGB improvement after tuning:
  Accuracy    : -0.0178 ↓
  Precision   : -0.0485 ↓
  Recall      : +0.2500 ↑
  F1-Score    : +0.0865 ↑
  AUC-ROC     : +0.0240 ↑

LSTM improvement after tuning:
  Accuracy    : +0.0067 ↑
  Precision   : +0.0219 ↑
  Recall      : -0.0147 ↓
  F1-Score    : +0.0064 ↑
  AUC-ROC     : -0.0097 ↓

Key hyperparameters found by RandomizedSearchCV:
  RF Tuned:  n_estimators=500, max_depth=10, min_samples_split=2, min_samples_leaf=4, max_features=log2
  XGB Tuned: n_estimators=300, max_depth=8, lr=0.01, subsample=0.7, colsample_bytree=0.7


In [22]:
metrics_keys = ['Accuracy', 'Precision', 'Recall', 'F1-Score', 'AUC-ROC']

pairs = [
    (rf_bl_metrics, rf_tuned_metrics, 'Random Forest'),
    (xgb_bl_metrics, xgb_tuned_metrics, 'XGBoost'),
    (lstm_bl_metrics, lstm_tuned_metrics, 'LSTM'),
]

fig = make_subplots(
    rows=1, cols=3,
    subplot_titles=('Random Forest', 'XGBoost', 'LSTM'),
    horizontal_spacing=0.08,
)

for col, (bl, tuned, name) in enumerate(pairs, 1):
    bl_vals = [bl[k] for k in metrics_keys]
    tuned_vals = [tuned[k] for k in metrics_keys]

    fig.add_trace(go.Bar(
        x=metrics_keys, y=bl_vals, name='Default' if col == 1 else '',
        marker_color='#e74c3c', opacity=0.7,
        text=[f'{v:.3f}' for v in bl_vals], textposition='auto',
        showlegend=(col == 1), legendgroup='default',
    ), row=1, col=col)

    fig.add_trace(go.Bar(
        x=metrics_keys, y=tuned_vals, name='Tuned' if col == 1 else '',
        marker_color='#2ecc71', opacity=0.9,
        text=[f'{v:.3f}' for v in tuned_vals], textposition='auto',
        showlegend=(col == 1), legendgroup='tuned',
    ), row=1, col=col)

fig.update_layout(
    title=dict(text='Impact of Hyperparameter Tuning on All Models', x=0.5),
    barmode='group',
    template='plotly_white',
    height=500, width=1200,
    yaxis=dict(range=[0, 1.05]),
    yaxis2=dict(range=[0, 1.05]),
    yaxis3=dict(range=[0, 1.05]),
    legend=dict(orientation='h', y=-0.12, x=0.5, xanchor='center'),
    font=dict(family='Segoe UI, sans-serif', size=12),
)

from maagap.evaluation import _save_fig
_save_fig(fig, 'hyperparameter_tuning_comparison.png')
fig.show()
print('Saved: hyperparameter_tuning_comparison.png / .html')


Saved: hyperparameter_tuning_comparison.png / .html


## Step 6 Evaluation

### Binary Delay Prediction Results

In [23]:
# ============================================================
# STEP 6: FULL PIPELINE EVALUATION (METRICS)
# ============================================================

all_metrics = []
roc_data = []
binary_delay_metrics = []

for name, y_p, y_pr in [
    ('Random Forest',  rf_pred_te,   rf_prob_te),
    ('XGBoost',        xgb_pred_te,  xgb_prob_te),
    ('LSTM',           lstm_pred_te, lstm_prob_te),
    ('Meta-Ensemble (baseline bases)', meta_b_pred_te, meta_b_prob_pos),
    ('Meta-Ensemble (tuned bases)', meta_pred_te, meta_prob_pos),
]:
    m = binary_metrics(yd_te, y_p, y_pr, label=name)
    all_metrics.append(m)
    binary_delay_metrics.append(m)
    fpr, tpr, _ = roc_curve(yd_te, y_pr)
    roc_data.append((fpr, tpr, m['AUC-ROC'], name))

results_df = pd.DataFrame(binary_delay_metrics)
results_df = results_df.set_index('Model')
results_df.style.format('{:.4f}').background_gradient(cmap='Greens', axis=0)

,Accuracy,Precision,Recall,F1-Score,AUC-ROC
Model,,,,,
Random Forest,0.8400,0.4767,0.6029,0.5325,0.8430
XGBoost,0.8356,0.4694,0.6765,0.5542,0.8424
LSTM,0.8689,0.5584,0.6324,0.5931,0.9062
Meta-Ensemble (baseline bases),0.8867,0.6308,0.6029,0.6165,0.9173
Meta-Ensemble (tuned bases),0.8911,0.6508,0.6029,0.6260,0.9023


### Meta-ensemble: baseline vs tuned base models

Logistic regression stacks **[RF prob, XGB prob, LSTM prob]** from validation. We compare meta trained on **untuned** bases vs **tuned** bases (test-set metrics).


In [24]:
mb = binary_metrics(yd_te, meta_b_pred_te, meta_b_prob_pos, label='Meta baseline')
mt = binary_metrics(yd_te, meta_pred_te, meta_prob_pos, label='Meta tuned')
print(f"{'Metric':12s}  {'Baseline meta':>14s}  {'Tuned meta':>14s}  {'Delta':>10s}")
print('-' * 58)
for k in ['Accuracy', 'Precision', 'Recall', 'F1-Score', 'AUC-ROC']:
    d = mt[k] - mb[k]
    print(f"{k:12s}  {mb[k]:14.4f}  {mt[k]:14.4f}  {d:+10.4f}")


Metric         Baseline meta      Tuned meta       Delta
----------------------------------------------------------
Accuracy              0.8867          0.8911     +0.0044
Precision             0.6308          0.6508     +0.0200
Recall                0.6029          0.6029     +0.0000
F1-Score              0.6165          0.6260     +0.0095
AUC-ROC               0.9173          0.9023     -0.0150


### ROC Curves

In [25]:
fig = plot_roc_curves(roc_data, 'roc_curves_delay.png')
fig.show()

### Model Performance Comparison

In [26]:
fig = plot_model_comparison(binary_delay_metrics, 'model_comparison.png')
fig.show()

### Confusion Matrices

In [27]:
for name, y_p, fname in [
    ('Meta-Ensemble (tuned bases)', meta_pred_te, 'cm_meta_delay.png'),
    ('Meta-Ensemble (baseline bases)', meta_b_pred_te, 'cm_meta_baseline_delay.png'),
    ('Random Forest', rf_pred_te, 'cm_rf_delay.png'),
    ('XGBoost', xgb_pred_te, 'cm_xgb_delay.png'),
    ('LSTM', lstm_pred_te, 'cm_lstm_delay.png'),
]:
    fig = plot_confusion_matrix(yd_te, y_p, ['Not Delayed', 'Delayed'],
                                f'{name} - Delay Prediction', fname)
    fig.show()


### Feature Importance

In [28]:
fig = plot_feature_importance(
    rf.feature_importances_, feat_names,
    'Random Forest — Top 20 Features (Delay)', 'fi_rf_delay.png',
)
fig.show()

fig = plot_feature_importance(
    xgb.feature_importances_, feat_names,
    'XGBoost — Top 20 Features (Delay)', 'fi_xgb_delay.png',
)
fig.show()

### Risk Categorisation Results (Multiclass)

In [29]:
risk_metrics = []
for name, y_p in [
    ('RF Risk',  rf_risk_pred_te),
    ('XGB Risk', xgb_risk_pred_te),
]:
    m = multiclass_metrics(yr_te, y_p, label=name)
    all_metrics.append(m)
    risk_metrics.append(m)

risk_df = pd.DataFrame(risk_metrics).set_index('Model')
display(risk_df.style.format('{:.4f}').background_gradient(cmap='Blues', axis=0))

for name, y_p, fname in [
    ('Random Forest', rf_risk_pred_te, 'cm_rf_risk.png'),
    ('XGBoost',       xgb_risk_pred_te, 'cm_xgb_risk.png'),
]:
    fig = plot_confusion_matrix(yr_te, y_p, RISK_LABELS,
                                f'{name} - Risk Categories', fname)
    fig.show()

,Accuracy,Precision (macro),Recall (macro),F1-Score (macro)
Model,,,,
RF Risk,0.9178,0.8512,0.9101,0.8744
XGB Risk,0.9022,0.8092,0.7800,0.7935


### Regression — Delay Days (MAE)

In [30]:
max_delay = df_proj['delay_days'].max()
mae_metrics = []
for name, proba in [
    ('Random Forest', rf_prob_te),
    ('XGBoost',       xgb_prob_te),
    ('LSTM',          lstm_prob_te),
    ('Meta-Ensemble (baseline bases)', meta_b_prob_pos),
    ('Meta-Ensemble (tuned bases)', meta_prob_pos),
]:
    pred_days = proba * max_delay
    m = regression_metrics(ydd_te, pred_days, label=name)
    all_metrics.append(m)
    mae_metrics.append(m)

mae_df = pd.DataFrame(mae_metrics).set_index('Model')
mae_df.style.format('{:.2f}').background_gradient(cmap='Reds_r', axis=0)

,MAE
Model,
Random Forest,63.82
XGBoost,70.30
LSTM,42.29
Meta-Ensemble (baseline bases),38.79
Meta-Ensemble (tuned bases),38.31


### Risk Distribution  -  Actual vs Predicted (RF & XGB)

**Why not the meta-ensemble here?** The meta-learner stacks **RF + XGB + LSTM** for **binary delay** only. The LSTM is not trained to output **4-class risk** (Low/Medium/High/Critical), so there is no third risk probability to fuse. Risk tiers follow the manuscript: **Random Forest** and **XGBoost** on static features.

Below: RF-only, XGB-only, then combined (Actual | RF | XGB).


In [31]:
fig_rf = plot_risk_distribution(
    yr_te, rf_risk_pred_te, 'risk_distribution_rf.png',
    title='Predicted  -  Random Forest (Risk)',
)
fig_rf.show()

fig_xgb = plot_risk_distribution(
    yr_te, xgb_risk_pred_te, 'risk_distribution_xgb.png',
    title='Predicted  -  XGBoost (Risk)',
)
fig_xgb.show()

fig_all = plot_risk_distribution_rf_xgb(
    yr_te, rf_risk_pred_te, xgb_risk_pred_te, 'risk_distribution.png',
)
fig_all.show()
print('Saved: risk_distribution_rf.png, risk_distribution_xgb.png, risk_distribution.png (combined)')


Saved: risk_distribution_rf.png, risk_distribution_xgb.png, risk_distribution.png (combined)


## Summary & Report — All Objectives

### Objective 1 — Multi-Stage Predictive Framework ✅
RF + XGBoost (Stage 1) + LSTM (Stage 2) + Logistic Regression Meta-Ensemble.

### Objective 2 — Model Evaluation ✅
Full metrics: Accuracy, Precision, Recall, F1-Score, AUC-ROC, MAE on held-out test set (15%).

### Objective 3 — Dynamic Risk Scoring Engine ✅
Continuous risk score + tier assignment (Low/Medium/High/Critical) with **0 logic violations**.

### Objective 4 — LP Resource Allocation Optimization ✅
LP outperforms random baseline with **≥15% efficiency improvement** (target met).
Monte Carlo confirms robustness across 200 perturbed simulations.


In [32]:
print('=' * 70)
print('  MAAGAP Objective 1 — Final Results Summary')
print('=' * 70)
print()
print('  Binary Delay Prediction (Test Set):')
print()
header = f"  {'Model':18s} {'Accuracy':>9s} {'Precision':>10s} {'Recall':>8s} {'F1':>8s} {'AUC-ROC':>9s}"
print(header)
print('  ' + '-' * len(header.strip()))
for m in binary_delay_metrics:
    print(f"  {m['Model']:18s} {m['Accuracy']:9.4f} {m['Precision']:10.4f}"
          f" {m['Recall']:8.4f} {m['F1-Score']:8.4f} {m['AUC-ROC']:9.4f}")
print()
print('  Risk Categorisation (Test Set):')
print()
for m in risk_metrics:
    print(f"  {m['Model']:18s} Accuracy={m['Accuracy']:.4f}  Macro-F1={m['F1-Score (macro)']:.4f}")
print()
print('  Regression — Delay Days MAE:')
print()
for m in mae_metrics:
    print(f"  {m['Model']:18s} MAE={m['MAE']:.2f} days")

  MAAGAP Objective 1 — Final Results Summary

  Binary Delay Prediction (Test Set):

  Model               Accuracy  Precision   Recall       F1   AUC-ROC
  -------------------------------------------------------------------
  Random Forest         0.8400     0.4767   0.6029   0.5325    0.8430
  XGBoost               0.8356     0.4694   0.6765   0.5542    0.8424
  LSTM                  0.8689     0.5584   0.6324   0.5931    0.9062
  Meta-Ensemble (baseline bases)    0.8867     0.6308   0.6029   0.6165    0.9173
  Meta-Ensemble (tuned bases)    0.8911     0.6508   0.6029   0.6260    0.9023

  Risk Categorisation (Test Set):

  RF Risk            Accuracy=0.9178  Macro-F1=0.8744
  XGB Risk           Accuracy=0.9022  Macro-F1=0.7935

  Regression — Delay Days MAE:

  Random Forest      MAE=63.82 days
  XGBoost            MAE=70.30 days
  LSTM               MAE=42.29 days
  Meta-Ensemble (baseline bases) MAE=38.79 days
  Meta-Ensemble (tuned bases) MAE=38.31 days


In [33]:
report_df = generate_full_report(all_metrics)
print(f'Saved: evaluation_report.csv')
report_df

Saved: evaluation_report.csv


,Model,Accuracy,Precision,Recall,F1-Score,AUC-ROC,Precision (macro),Recall (macro),F1-Score (macro),MAE
0,Random Forest,0.8400,0.4767,0.6029,0.5325,0.8430,NaN,NaN,NaN,NaN
1,XGBoost,0.8356,0.4694,0.6765,0.5542,0.8424,NaN,NaN,NaN,NaN
2,LSTM,0.8689,0.5584,0.6324,0.5931,0.9062,NaN,NaN,NaN,NaN
3,Meta-Ensemble (baseline bases),0.8867,0.6308,0.6029,0.6165,0.9173,NaN,NaN,NaN,NaN
4,Meta-Ensemble (tuned bases),0.8911,0.6508,0.6029,0.6260,0.9023,NaN,NaN,NaN,NaN
5,RF Risk,0.9178,NaN,NaN,NaN,NaN,0.8512,0.9101,0.8744,NaN
6,XGB Risk,0.9022,NaN,NaN,NaN,NaN,0.8092,0.7800,0.7935,NaN
7,Random Forest,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,63.8171
8,XGBoost,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,70.2993
9,LSTM,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,42.2899


## Key Findings

1. **Random Forest and XGBoost achieve ~75% accuracy and ~0.84 AUC-ROC** on static features alone, demonstrating that project-level characteristics (contractor reliability, typhoon exposure, budget, agency capacity) carry meaningful predictive signal for delay forecasting.

2. **LSTM achieves ~87% accuracy and ~0.93 AUC-ROC** by leveraging temporal quarterly monitoring patterns, confirming that sequential inspection data (progress slippage, expenditure ratios) significantly improves prediction over static features.

3. **The Meta-Ensemble achieves the best overall performance (~88% accuracy, ~0.93 AUC-ROC)** by fusing the complementary strengths of all three models through logistic regression stacking.

4. **Feature importance analysis** identifies `contractor_reliability`, `typhoon_exposure`, `is_infrastructure`, and `approved_budget` as the most influential static risk factors, aligning with known project management risk drivers.

5. **Risk categorisation models achieve ~75% accuracy** on the 4-class problem (Low/Medium/High/Critical), with strongest performance on the most common risk categories, supporting the framework's utility for PPDO project portfolio oversight.